File: aux_scripts/get_stats.rb
#!/usr/bin/env ruby

require 'expcalc'
require 'optparse'

def get_stats(data)
      stats = Hash.new(0)
      stats[:average] = data.sum().fdiv(data.size)
      sum_devs = data.sum{|element| (element - stats[:avg]) ** 2}
      stats[:variance] = sum_devs.fdiv(data.size)
      stats[:standardDeviation] = stats[:variance] ** 0.5
      stats[:max] = data.max
      stats[:min] = data.min

      stats[:count] = data.size
      data.each do |value|
        stats[:countNonZero] += 1 if value != 0
      end

      stats[:q1] = data.get_quantiles(0.25)
      stats[:median] = data.get_quantiles(0.5)
      stats[:q3] = data.get_quantiles(0.75)
      return stats
end


def report_stats(stats)
  report_stats = []
  report_stats << ['Elements', stats[:count]]
  report_stats << ['Elements Non Zero', stats[:countNonZero]]
  report_stats << ['Non Zero Density', stats[:countNonZero].fdiv(stats[:count])]
  report_stats << ['Max', stats[:max]]
  report_stats << ['Min', stats[:min]]
  report_stats << ['Average', stats[:average]]
  report_stats << ['Variance', stats[:variance]]
  report_stats << ['Standard Deviation', stats[:standardDeviation]]
  report_stats << ['Q1', stats[:q1]]
  report_stats << ['Median', stats[:median]]
  report_stats << ['Q3', stats[:q3]]
  return report_stats
end

def extract_data(lst_file)
    data = []

    File.open(lst_file,"r").each do |line|
    line.chomp!
    data.append(line.to_f)
    end

    return data
end


########################### OPTPARSE ########################
#############################################################

options = {}

OptionParser.new do |opts|

  options[:data_file] = nil
  opts.on("-d","-data_file DATA", "The path to the DATA file") do |data_file|
    options[:data_file] = data_file
  end
  
end.parse!


####################################################################

path2data=options[:data_file]

data = extract_data(path2data)
stats = get_stats(data)
report_stats(stats).each do |stat|
    puts stat.join("\t")
end

File: scripts/aux_scripts/get_stats.py
#!/usr/bin/env python

import numpy as np
import argparse

def get_stats(data):
	stats = {}
	stats['average'] = np.mean(data)
	stats['variance'] = np.var(data)
	stats['standardDeviation'] = np.std(data)
	stats['max'] = np.max(data)
	stats['min'] = np.min(data)
	stats['count'] = len(data)
	stats['countNonZero'] = np.count_nonzero(data)
	stats['q1'], stats['median'], stats['q3'] = np.percentile(data, [25, 50, 75])
	return stats

def report_stats(stats):
	report_stats = []
	report_stats.append(['Elements', stats['count']])
	report_stats.append(['Elements Non Zero', stats['countNonZero']])
	report_stats.append(['Non Zero Density', stats['countNonZero'] / stats['count']])
	report_stats.append(['Max', stats['max']])
	report_stats.append(['Min', stats['min']])
	report_stats.append(['Average', stats['average']])
	report_stats.append(['Variance', stats['variance']])
	report_stats.append(['Standard Deviation', stats['standardDeviation']])
	report_stats.append(['Q1', stats['q1']])
	report_stats.append(['Median', stats['median']])
	report_stats.append(['Q3', stats['q3']])
	return report_stats

def extract_data(lst_file):
	with open(lst_file, 'r') as f:
		data = [float(line.strip()) for line in f]
	return data


########################### OPTPARSE ########################
#############################################################

parser = argparse.ArgumentParser(description='Calculate statistics of data in a file')
parser.add_argument('-d', '--data_file', help='The path to the data file', required=True)
options = parser.parse_args()

####################################################################

path2data = options.data_file

data = extract_data(path2data)
stats = get_stats(data)
for stat in report_stats(stats):
	print('\t'.join(map(str, stat)))